### PR TITLE
Implement OpenMP built-in reduction construct

### PR DESCRIPTION
**Pull request by mfournial**
* [x] Implement the `__kmpc_reduce`, `__kmpc_end_reduce`, `__kmpc_reduce_nowait`, `__kmpc_reduce_nowait` functions, depending on what's most important.
* [x] Make simpler reduce cases (e.g. avg, add) (see `reduce_examples`), including one with a custom function.
* [x] Tackle Clang crashing bug when compiling `simple_reduce.cpp` with the Faasm toolchain... see below:
Native ltrace:
```
% time     seconds  usecs/call     calls      function
------ ----------- ----------- --------- --------------------
 80.67    0.861549      861549         1 __kmpc_fork_call
 17.49    0.186813       46703         4 __kmpc_end_reduce
  0.93    0.009963        1992         5 __kmpc_global_thread_num
  0.36    0.003809         952         4 __kmpc_barrier
  0.15    0.001568         392         4 __kmpc_for_static_fini
  0.13    0.001392         348         4 omp_get_thread_num
  0.13    0.001392         348         4 __kmpc_reduce
  0.12    0.001313         328         4 __kmpc_for_static_init_4
  0.01    0.000145         145         1 __kmpc_push_num_threads
------ ----------- ----------- --------- --------------------
100.00    1.067944                    31 total
```

But this currently breaks clang:
```
2.	Code generation
toolchain/bin/clang-9(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x2a)[0x55bd07332097]
toolchain/bin/clang-9(_ZN4llvm3sys17RunSignalHandlersEv+0x30)[0x55bd0733096b]
toolchain/bin/clang-9(+0xe6ca26)[0x55bd07330a26]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f046ccba890]
[0x55bd0bdf4e10]
clang-9: error: unable to execute command: Segmentation fault (core dumped)
```

### CODE REVIEW

File: lib-faasmp/z_Linux_asm.S [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// It is therefore only named and not used because we do not care about
+// compatibility with any runtime
+	.hidden	.gomp_critical_user_.reduction.var # @.gomp_critical_user_.reduction.var
+	.type	.gomp_critical_user_.reduction.var,@object
+	.section	.data..gomp_critical_user_.reduction.var,"",@
+	.globl	.gomp_critical_user_.reduction.var
+	.p2align	4
+.gomp_critical_user_.reduction.var:
+	.int64	42                      # 0x2a
+	.size	.gomp_critical_user_.reduction.var, 8
```
Comment (Shillaker) [inline lib-faasmp/z_Linux_asm.S]: :sunglasses:

File: test/wasm/test_openmp.cpp [hunk 2]
```diff
@@ -46,4 +46,10 @@ namespace tests {
         message::Message msg = util::messageFactory("omp", "stack_debug");
         execFunction(msg);
     }
+
+    TEST_CASE("Test simple reduction function", "[wasm]") {
+        cleanSystem();
+        message::Message msg = util::messageFactory("omp", "simple_reduce");
+        execFunction(msg);
+    }
```
Comment (Shillaker) [inline test/wasm/test_openmp.cpp]: Would be nice to add some more of the `omp` functions as tests to get a broader coverage.

Comment (mfournial) [reply test/wasm/test_openmp.cpp]: This is done, there is a lot of constructs tested now. I adapted the averaging and the integral code I found online to now test the return values

**Comment by Shillaker** [comment]
LGTM, will merge tomorrow

**Pull request merged**